README.rst
.. highlight:: shell

===============================
Drone dectector
===============================

.. image:: https://img.shields.io/github/v/tag/malapert/drone_dectector
.. image:: https://img.shields.io/github/v/release/malapert/drone_dectector?include_prereleases

.. image https://img.shields.io/github/downloads/malapert/drone_dectector/total
.. image https://img.shields.io/github/issues-raw/malapert/drone_dectector
.. image https://img.shields.io/github/issues-pr-raw/malapert/drone_dectector
.. image:: https://img.shields.io/badge/Maintained%3F-yes-green.svg
   :target: https://github.com/malapert/drone_dectector/graphs/commit-activity
.. image https://img.shields.io/github/license/malapert/drone_dectector
.. image https://img.shields.io/github/forks/malapert/drone_dectector?style=social


Drone detector

From sources
------------

The sources for Drone dectector can be downloaded from the `Github repo`_.

You can either clone the public repository:

.. code-block:: console

    $ git clone git://github.com/malapert/drone_dectector

Or download the `tarball`_:

.. code-block:: console

    $ curl -OJL https://github.com/malapert/drone_dectector/tarball/master

Once you have a copy of the source, you can install it with:

.. code-block:: console

    $ tar zxvf drone_dectector-0.1.dev0.tar.gz
    $ cd drone_dectector-0.1.dev0
    $ make  # install in the system root
    $ make user # or Install for non-root usage


.. _Github repo: https://github.com/malapert/drone_dectector
.. _tarball: https://github.com/malapert/drone_dectector/tarball/master



Development
-----------

.. code-block:: console

        $ git clone https://github.com/malapert/drone_dectector
        $ cd drone_dectector
        $ make prepare-dev
        $ source .drone_dectector
        $ make install-dev


To get more information about the preconfigured tasks:

.. code-block:: console

        $ make help

Usage
-----

To use Drone dectector in a project::

.. code-block:: console

        usage: drone_dectector [-h] [-v] [--level {INFO,DEBUG,WARNING,ERROR,CRITICAL,TRACE}] [--contour_area CONTOUR_AREA]
                            [--treshold TRESHOLD]
                            {camera,video} ...

        Drone detector

        positional arguments:
        {camera,video}

        options:
        -h, --help            show this help message and exit
        -v, --version         show program's version number and exit
        --level {INFO,DEBUG,WARNING,ERROR,CRITICAL,TRACE} set Level log (default: INFO)
        --contour_area CONTOUR_AREA set the box area from which the contour is drawn (default: 50)
        --treshold TRESHOLD   Threshold (from 0 to 255) for which we consider the images are different between the two last images (>20 /
                                255) (default: 20)


Some examples :

    * drone_dectector camera
    * drone_dectector --contour_area 80 camera
    * drone_dectector --contour_area 80 --level DEBUG camera
    * drone_dectector video --file /home/malapert/Téléchargements/test.mp4

Run tests
---------

.. code-block:: console

        $make tests



Author
------
👤 **Jean-Christophe Malapert**



🤝 Contributing
---------------
Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/malapert/drone_dectector/issues). You can also take a look at the [contributing guide](https://github.com/malapert/drone_dectector/blob/master/CONTRIBUTING.rst)


📝 License
----------
This project is [GNU General Public License v3](https://github.com/malapert/drone_dectector/blob/master/LICENSE) licensed.


CHANGELOG
### First version (HEAD -> master)
>Wed, 4 Jan 2023 01:32:15 +0100

>Author: Jean-Christophe Malapert (Jean-christophe.malapert@cnes.fr)

>Commiter: Jean-Christophe Malapert (Jean-christophe.malapert@cnes.fr)






drone_dectector/__main__.py
# Drone dectector - Drone detector
# Copyright (C) 2023 - Malapert
#
# This file is part of Drone dectector.
#
# Drone dectector is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Drone dectector is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Drone dectector.  If not, see <https://www.gnu.org/licenses/>.
"""Main program."""
import logging
import argparse
import signal
import sys
from drone_dectector import __author__
from drone_dectector import __copyright__
from drone_dectector import __description__
from drone_dectector import __version__
from .drone_dectector import DroneDectectorLib


class SmartFormatter(argparse.HelpFormatter):
    """Smart formatter for argparse - The lines are split for long text"""

    def _split_lines(self, text, width):
        if text.startswith("R|"):
            return text[2:].splitlines()
        # this is the RawTextHelpFormatter._split_lines
        return argparse.HelpFormatter._split_lines( # pylint: disable=protected-access
            self, text, width
        )


class SigintHandler:  # pylint: disable=too-few-public-methods
    """Handles the signal"""

    def __init__(self):
        self.SIGINT = False   # pylint: disable=invalid-name

    def signal_handler(self, sig: int, frame):
        """Trap the signal

        Args:
            sig (int): the signal number
            frame: the current stack frame
        """
        # pylint: disable=unused-argument
        logging.error("You pressed Ctrl+C")
        self.SIGINT = True
        sys.exit(2)


def str2bool(string_to_test: str) -> bool:
    """Checks if a given string is a boolean

    Args:
        string_to_test (str): string to test

    Returns:
        bool: True when the string is a boolean otherwise False
    """
    return string_to_test.lower() in ("yes", "true", "True", "t", "1")


def parse_cli() -> argparse.Namespace:
    """Parse command line inputs.

    Returns
    -------
    argparse.Namespace
        Command line options
    """
    parser = argparse.ArgumentParser(
        description=__description__,
        formatter_class=SmartFormatter,
        epilog=__author__ + " - " + __copyright__,
    )
    parser.add_argument(
        "-v", "--version", action="version", version="%(prog)s " + __version__
    )  

    parser.add_argument(
        "--level",
        choices=[
            "INFO",
            "DEBUG",
            "WARNING",
            "ERROR",
            "CRITICAL",
            "TRACE",
        ],
        default="INFO",
        help="set Level log (default: %(default)s)",
    )
    
    parser.add_argument(
        "--contour_area",
        default=50,
        type = int,
        help="set the box area from which the contour is drawn (default: %(default)s)",
    )  
    
    parser.add_argument(
        "--treshold",
        default=20,
        type = int,
        help="Threshold (from 0 to 255) for which we consider the images are different between the two last images (>20 / 255) (default: %(default)s)",
    )       
    
    subparser = parser.add_subparsers()
    camera = subparser.add_parser('camera')
    video = subparser.add_parser('video')
    video.add_argument(
        '--file',
        required= True,
        type = str,
        help="Path to the video"
    )      

    return parser.parse_args()


def run():
    """Main function that instanciates the library."""
    logger = logging.getLogger(__name__)
    handler = SigintHandler()
    signal.signal(signal.SIGINT, handler.signal_handler)
    try:
        options_cli = parse_cli()
        drone_dectector = DroneDectectorLib(
            treshold = options_cli.treshold,
            contour_area = options_cli.contour_area,
            file=options_cli.file if hasattr(options_cli, 'file') else 0,
            level=options_cli.level,
        )
        drone_dectector.detect()
        sys.exit(0)
    except Exception as error:  # pylint: disable=broad-except
        logging.exception(error)
        sys.exit(1)


if __name__ == "__main__":
    # execute only if run as a script
    run()
